include/net/socket.h

```diff
@@ -390,6 +390,24 @@ static inline char *inet_ntop(sa_family_t family, const void *src, char *dst,
 #define EAI_SERVICE DNS_EAI_SERVICE
 #endif /* defined(CONFIG_NET_SOCKETS_POSIX_NAMES) */
 
+struct net_socket_register {
+	int family;
```

Comment:
We have family check in sockets.c so the value is not obsolete

Reply:
which is redundant

Reply:
Please elaborate this. The check is there because of TLS so that we do not have to register two handlers (which is possible of course if we want to).

Reply:
the functions check() (or any better name) will also verify that it supports the given family, so to me it seems redundant to check it before calling the check function. 

Also, if some handler wants to support more than 2+ family, it it will need to register 2+ times (once for each family it supports). Not great either. Or like you did in tls, it registers itself as AF_UNSPEC to avoid that, but again there is then no point of having this attribute here.

Reply:
> check is there because of TLS

I'd say that TLS should stay as it is, as a tribute to layering violation it introduces ;-). Then, we need to dispatch only on family, and check function isn't needed.

Reply:
> the functions check() (or any better name) will also verify that it supports the given family, so to me it seems redundant to check it before calling the check function.

Except that calling a function by pointer is several times slower than just comparing with a static number. So, per above, I think that the "check" function should go, not the "family" field.

> Also, if some handler wants to support more than 2+ family,

This isn't a too often/realistic case, when it happens (AF_INET/AF_INET6, period?), adding 2 entries indeed should be a good way.

Reply:
> the functions check() (or any better name) will also verify that it supports the given family, so to me it seems redundant to check it before calling the check function.

Sure that is one option. I added the family check to sockets.c so that we do not need to call the check function for those families that we definitely are not able to support.

Reply:
@pfalcon 
> Except that calling a function by pointer is several times slower than just comparing with a static number. So, per above, I think that the "check" function should go, not the "family" field.

And how proto is checked then?

Reply:
In my original code, I actually had both family, type and proto checks, but it can get really messy real fast (especially when doing unit test). There can be many combinations of these so decided to keep it simple and only check the family as probably there are only limited number of these.

Reply:
@tbursztyka 

> And how proto is checked then?

Proto and anything else is checked by the module which implements a particular socket family. That's a pretty clean architecture - we have one self-contained module which implements entire socket family. This seems to work so far, the only exceptions is AF_INET and AF_INET6 implemented by one module (that's ok). And TLS, which as an exception at all, not a part POSIX/BSD/Linux socket API.

Note that I'm usually all for flexibility, but here IMHO we should set guidelines to avoid too much freedom which may lead to messy structure of modules. At least, I'd suggest to start with dispatching just on family, we can add more fine-grained dispatch if really warranted.